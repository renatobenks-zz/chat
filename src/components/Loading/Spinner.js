// @flow

import styled from 'styled-components';

const Spinner = styled.div`
  width: 40px;
  height: 40px;
  background-color: ${props => props.color || props.theme.palette.white};
  animation: sk-rotateplane 1.2s infinite ease-in-out;

  @keyframes sk-rotateplane {
    0% {
      transform: perspective(120px) rotateX(0deg) rotateY(0deg);
    }

    50% {
      transform: perspective(120px) rotateX(-180.1deg) rotateY(0deg);
    }

    100% {
      transform: perspective(120px) rotateX(-180deg) rotateY(-179.9deg);
    }
  }
`;

export default Spinner;
