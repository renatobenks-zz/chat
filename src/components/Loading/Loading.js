// @flow

import * as React from 'react';
import styled, { css } from 'styled-components';

import Spinner from './Spinner';

export const TYPES = {
  SPINNER: 'spinner',
};

type Type = $Values<typeof TYPES>;

const SPINNERS = {
  [TYPES.SPINNER]: Spinner,
};

const Wrapper = styled.section`
  display: flex;
  padding: 1rem;
  flex-flow: column wrap;
  align-items: center;
  justify-content: center;
  background: ${props => props.background || props.theme.palette.primary};
  ${props =>
    props.full &&
    css`
      width: 100%;
      height: 100%;
      position: absolute;
      top: 0;
      left: 0;
      z-index: 99;
    `};
`;

type Props = {
  visible?: boolean,
  full?: boolean,
  type?: Type,
  color?: string,
  background?: string,
  children?: React.Element<any>,
};

const Loading = ({
  visible, full, type, background, children, ...props 
}: Props) => {
  if (!visible) return null;

  const Spinner = (type && SPINNERS[type]) || SPINNERS[TYPES.SPINNER];

  return (
    <Wrapper full={full} background={background}>
      <Spinner {...props} />
      {children}
    </Wrapper>
  );
};

Loading.defaultProps = {
  visible: true,
  full: false,
};

export default Loading;
