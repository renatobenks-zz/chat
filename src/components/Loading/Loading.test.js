import React from 'react';

import { render } from '../../../test/helpers';

import Loading, { TYPES } from './Loading';

const renderer = component => render(component, { withTheme: true });

it('loading as default', () => {
  expect(renderer(<Loading />)).toMatchSnapshot();
});

it('full loading', () => {
  expect(renderer(<Loading full />)).toMatchSnapshot();
});

it('spinner loading type', () => {
  expect(renderer(<Loading type={TYPES.SPINNER} />)).toMatchSnapshot();
});

it('not visible loading', () => {
  expect(renderer(<Loading visible={false} />)).toMatchSnapshot();
});
