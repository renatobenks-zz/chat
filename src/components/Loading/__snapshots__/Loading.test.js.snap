// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full loading 1`] = `
<section
  className="sc-bwzfXH gVSmEA"
>
  <div
    className="sc-bdVaJa dAIsho"
  />
</section>
`;

exports[`loading as default 1`] = `
<section
  className="sc-bwzfXH hgQgvr"
>
  <div
    className="sc-bdVaJa dAIsho"
  />
</section>
`;

exports[`loading with no background 1`] = `
<section
  className="sc-bwzfXH fTZzux"
>
  <div
    className="sc-bdVaJa dAIsho"
  />
</section>
`;

exports[`loading with no wrapper 1`] = `
<div
  className="sc-bdVaJa dAIsho"
/>
`;

exports[`not visible loading 1`] = `null`;

exports[`spinner loading type 1`] = `
<section
  className="sc-bwzfXH hgQgvr"
>
  <div
    className="sc-bdVaJa dAIsho"
  />
</section>
`;
